### PR TITLE
Save node_modules!

### PR DESCRIPTION
**Pull request by theKashey**
Very rare, but sometimes you have to override some global. 
And proxyquire will wipe require.cache to perform it.

Some time ago you `secure` native node.js modules from this, but, to say the truth, you should dont touch any non-project files.
As long you are not testing them, and not mocking - keep them alive.

And second important thing, from React world, you cannot mix together React components from different versions of React. Not versions - generations.

So, sometimes node_modules must be secured. And they can, cos they are third party libraries.

Solution - .onlyForProjectFiles and .forAllFiles.
PS: no changes in .ts file.

### CODE REVIEW

**Comment by bendrucker** [comment]
Tests are failing. More importantly, I can't say I understand what's happening here. I'd prefer to start with a failing test where you simulate a broken library.

File: lib/proxyquire.js [hunk 1]
```diff
@@ -76,7 +76,7 @@ Proxyquire.prototype.callThru = function () {
 };
 
 /**
- * Will make proxyquire remove the requested modules from the `require.cache` in order to force 
+ * Will make proxyquire remove the requested modules from the `require.cache` in order to force
```
Comment (bendrucker) [inline lib/proxyquire.js]: fix the whitespace

File: lib/proxyquire.js [hunk 2]
```diff
@@ -218,28 +245,42 @@ Proxyquire.prototype._disableCache = function(module, path) {
...
-    if (/\.node$/.test(id)) {
-      require.cache[id] = cache[id];
+  function wipeCache(cache, restoredCache, target) {
+    for (var id in cache) {
+      var keep = false;
+      // Keep native modules (i.e. `.node` files).
+      // Otherwise, Node.js would throw a “Module did not self-register”
+      // error upon requiring it a second time.
+      // See https://github.com/nodejs/node/issues/5016.
+      if (/\.node$/.test(id)) {
```
Comment (bendrucker) [inline lib/proxyquire.js]: this is gross, just use an `if (a || b)` and set the cache key in there

**Comment by thlorenz** [comment]
As is this PR isn't acceptable .. agree w/ @bendrucker  there.

Whitespace changes need to be removed and please explain with a simple example what you're trying to achieve.
I'm also very worried that this breaks a bunch of users as it clearly changes some behavior. Most likely that's why those tests are failing.

So at this point I'm a big :-1: on this change, but will leave it up to @bendrucker to make the final decision assuming things get cleaned up first.

**Comment by theKashey** [comment]
@thlorenz - this PR did not break anything. Default behavior is untouched.
But let me explain how it works. Any why.

Lets open test/proxyquire-global.js. And it will next open some modules in samples directory. 
How many other modules from node_modules are opened at this point? 
Something around 50 - mocha family, lodash, and so on.
As long this is external modules and you using them, not them using you - you will not mock anything inside them.
So, there is no reason to wipe them from global cache, if you want to override you own runtime global.
You should work only with your own project files.
PS: And yes, due to absolutely clear cache some tests can took ~1m to be executed. Or 10ms with this patch.

Secondary target is almost same.
Lets imaging some code.
```javascript
import Button from 'react-components/Button'
...
const component = shallow(proxyquire("somereactcomponent",{override some global}))
expect(component.find(Button)).to.be.present;
```
`Button` will be _normal_ React component. With some magic react-uuid in internals.
`component` - will also be normal. But before you will get it - you will wipe React from cache. As result `component` will include different magic uuid.
And they will not work together.

So yes - sometime you have to override node_modules, but in mosts cases mock, not wipe.

**Comment by bendrucker** [comment]
Is there any other case where this is relevant outside of React? If so I'm 👎  on a feature specifically to support its idiosyncrasies.

**Comment by theKashey** [comment]
So I just indicate 2 edge cases:
 - when is better not to wipe EVERYTHING. Just to speed up tests. Just cos you shall not wipe external things, but only things belonging to your project
 - when you just cannot wipe something. React is just a example. Generating some randoms during startup is a common things in computer science.

So only two words - better, and have-to.

**Comment by bendrucker** [comment]
Sorry, not convinced. All code is work to maintain and support. Not saying this isn't valid for you but we aren't going to accept it at the moment.
